protos/feast/core/CoreService.proto

```diff
@@ -22,139 +22,49 @@ option go_package = "github.com/gojek/feast/protos/generated/go/feast/core";
-  // Register a new entity to the metadata store, or update an existing entity.
-  // If any validation errors occur, only the first encountered error will be returned.
-  rpc ApplyEntity(feast.specs.EntitySpec) returns (CoreServiceTypes.ApplyEntityResponse);
-  // Retrieve the topic and topic configuration (such as the URI of the message brokers)
-  // for the client to write the feature values of an entity.
-  rpc GetTopic(CoreServiceTypes.GetTopicRequest) returns (CoreServiceTypes.GetTopicResponse);
-  // Request a signed URL where a Feast client can upload a feature values file for an import job.
-  // The signed URL will be valid by default for 5 minutes during which the client can start
-  // uploading the feature values file. As of 2019-06-28, only CSV and JSON files are supported,
-  // and the upload must complete in one PUT request i.e. resumable upload is not supported.
-  rpc GetUploadUrl(CoreServiceTypes.GetUploadUrlRequest) returns (CoreServiceTypes.GetUploadUrlResponse);
+    // Retrieve feature set details for a specific feature set
+    rpc GetFeatureSet (CoreServiceTypes.GetFeatureSetRequest) returns (CoreServiceTypes.GetFeatureSetResponse);
+
+    // Idempotent creation of feature set. Will not create a new feature set if schema does not change
+    rpc ApplyFeatureSet (CoreServiceTypes.ApplyFeatureSetRequest) returns (CoreServiceTypes.ApplyFeatureSetResponse);
 }
 
 message CoreServiceTypes {
-  message GetEntitiesRequest {
-    repeated string ids = 1;
-  }
-  message GetEntitiesResponse {
-    repeated feast.specs.EntitySpec entities = 1;
-  }
-  message ListEntitiesResponse {
-    repeated feast.specs.EntitySpec entities = 1;
-  }
-  // Feature retrieval
-  message GetFeaturesRequest {
-    repeated string ids = 1;
-  }
-  message GetFeaturesResponse {
-    repeated feast.specs.FeatureSpec features = 1;
-  }
-  message ListFeaturesResponse {
-    repeated feast.specs.FeatureSpec features = 1;
-  }
-  // Storage spec retrieval
-  message GetStorageRequest {
-    repeated string ids = 1;
-  }
-  message GetStorageResponse {
-    repeated feast.specs.StorageSpec storageSpecs = 1;
-  }
-  message ListStorageResponse {
-    repeated feast.specs.StorageSpec storageSpecs = 1;
-  }
-  // Entity registration response
-  message ApplyEntityResponse {
-    string entityName = 1;
-  }
-  // Feature registration response
-  message ApplyFeatureResponse {
-    string featureId = 1;
-  }
-  // Feature group registration response
-  message ApplyFeatureGroupResponse {
-    string featureGroupId = 1;
-  }
-  message GetUploadUrlRequest {
-    enum FileType {
-      CSV = 0;
-      JSON = 1;
+    // Retrieves details for all versions of a specific feature set
+    message GetFeatureSetRequest {
+        message FeatureSet {
+            string name = 1;
+
+            // (Optional) Only return details for a specific version
+            int32 version = 2;
+        }
+    }
+
+    message GetFeatureSetResponse {
```

Comment:
method is getFeatureSet but you get a repeated value? Seems unintuitive for the user

Reply:
That is right. The user gets a list of feature set versions for a single feature set name. Each feature set version is an item in that list.

Reply:
I think that the method name should hint towards the return type